allows more flexible timelion queries

ppisljar opened the pull request:
fixes #10907

allows more flexible usage of whitespace in timelion queries. 

this query now executes well:

```
.es     (    404  )   .yaxis  (   min   :   10   ) ;                 .es(*)      .yaxis    (    2 ,   min   =   10   )   ,  .es( 500)  
```

adds the option to use `;` instead of `,` but makes one of the separators obligatory (just white space is not allowed any more)

epixa commented:
Not support whitespace anymore is going to be a problem for existing saved expressions, no?  Unless there was some sort of in-place shim to support whitespace, this is a backwards compatibility break, and something we should consider an upgrade path for for 6.0.

Why is this a blocker, by the way?

cjcenizal commented:
@epixa This is a blocker for multi-line expressions (https://github.com/elastic/kibana/issues/8881), because multi-line expressions require whitespace. However, whitespace is currently interpreted as a delimiter between expressions. So, we need to remove this interpretation of whitespace, to allow for its use strictly as an organizational tool.

This is definitely a BWC break (as noted in https://github.com/elastic/kibana/issues/10907), but the use of whitespace as a delimiter isn't documented anywhere, so it's possible that not many people use it for that purpose. My suggestion is to display a warning to the user about the change and suggest using semicolons or commas to delimit expressions as a fix. We can delay to 6.0 if you want.

epixa commented:
Assuming I understand you correctly, I think that's a good plan.  Show deprecation notices for missing commas in 5.4, and remove their support entirely in 6.0.

If that complicates this feature, then just ship this feature in 6.0.

ppisljar commented:
might be complicated to add this warning. i could catch the error thrown by parser and in case of `
Error
Timelion: Error: in cell #1: metric.split is not a function` show the deprecation warning.
but i have no way of telling if this same error is also thrown in some other scenarios.

what do you think ? good enough ?

epixa commented:
@ppisljar I don't understand why we'd need to catch the error.  We would just need to identify when a space is used between two expressions and show a deprecation notice.  This PR would then only go into 6.0.

cjcenizal commented:
Correct if I'm wrong @ppisljar, but I think the reason we need to catch the error is because we don't know when a space is used between two expressions, or between two functions in a single expression, or in a query to ES. So we need to rely on Peg's parsing of the expressions to know if an expression is invalid (possibly due to our BWC breaking change).

If this understanding is correct, then I think we should show our deprecation notice when the error is thrown. It's OK if this error is thrown and we show the notice, even if the error is thrown due to some reason other than this change. We can play with the wording of the notice so that it still makes sense . But this error needs to be reliably thrown in all scenarios where this change creates an invalid expression or changes the meaning of an expression.

cjcenizal commented:
@ppisljar Can you add tests for this to verify which expressions are accepted and which throw errors?

* whitespace used between two expressions
* whitespace between two functions in a single expression
* whitespace in a query to ES
* semicolon-delimited expressions
* colon-delimited expressions
* any other variations you can think of

ppisljar commented:
@cjcenizal thats correct, regarding the parsing. I'll start adding tests.

rashidkpc commented:
Get rid of the semi-colon support. It add complication and ambiguity. Part of the reason for getting rid of whitespace delineation was to remove ambiguity, no reason to re-introduce it.

rashidkpc commented:
The pull description says this executes correctly:

```
.es     (    404  )   .yaxis  (   min   :   10   ) ; 
```

If that is true, then this is a bug, colons are not an acceptable argument assignment character. Some functions implement them for the purpose of string splitting, but they are not part of the grammar and shouldn't be.

rashidkpc commented:
Given that the purpose of #10907 is to make possible multi-line expressions, this would be a good time to introduce newlines and carriage returns as acceptable whitespace characters into the grammar. See here: https://github.com/elastic/timelion/commit/5b21a86a9e41cc9bdefc70f2b7e9416173a39b74#diff-cad677e1524a089d5994bbffedf06119

ppisljar commented:
@rashidkpc the min:10 doesn't actually work, but it doesn't throw an error either. however its not related to this PR.

i added the semicolon just to see what are the opinions on it ... i would prefer a semicolon there if we introduce the multiline editor ... 

```
.es(*)
  .yaxis(1);

.es(404)
  .yaxis(2);
```
vs
```
.es(*)
  .yaxis(1),

.es(404)
  .yaxis(2),
```

rashidkpc commented:
The semi-colon has no additional function over the comma. We should be making the expressions less ambiguous, not more.

Also, in your previous example, if that trailing comma is required, I'd consider that a bug.

cjcenizal commented:
@ppisljar If you feel strongly about the semi-colons, can we introduce and debate their merits in a separate PR?

rashidkpc commented:
@ppisljar This *should* throw an error if we're requiring comma delimiting as spaces aren't valid in unquoted strings. 

```
.yaxis  (   min   :   10   )
```

For example:

```
.es(
  q = html,
  index = logstash*
)

```

ppisljar commented:
i removed the semicolon support and added new line support

@rashidkpc 

yeah comma at the end is necessary in this version ... i have no idea how to make comma in the middle a must but comma at the end optional ... 

if you feel like doing this feel free to submit another PR and close this one.

rashidkpc commented:
@ppisljar I went ahead and made minimal the changes you'd need directly on this pull.

I also cleaned up the grammar a bit here, but chose not to commit those changes to this branch as, while the resulting parse is functionally equivalent, the diff is harder to read: https://github.com/rashidkpc/kibana/commit/c63ff9b870ba3b09bb010d0a0b7dcf0d0444e517

ppisljar commented:
@rashidkpc this now allows space as separator again (this works with no errors: `.es(404)   .es(200)` which shouldn't

rashidkpc commented:
@ppisljar That definitely does throw an error, the test is incorrect. 

The test is only running `parseSheet`, which serves to generate an AST to pass to the interpreter. Differentiating between datasources and chain-able functions is the responsibility of that interpreter, as there is no semantic difference between the two types. If there was, we wouldn't need comma separation as the grammar could observe the switch to the new chain at that point.

`.es(404) .es(200)` is valid in the grammar and implies that the output of `.es(404)` is being fed into `.es(200)` and that the two expressions are part of the same chain. The statement is proper grammatically but, as expected, throws an error at interpretation time.

Let's look at how, `.es(404) .es(200)` parses in the old, no commas required, grammar. I've removed the bits of the AST that assist with autocomplete to make this easier to read. As you can see, there are two chain objects here, instructing the parser to run `invokeChain()` on each, thus invoking each without any context.

```
 {
   "tree": [
      {
         "type": "chain",
         "chain": [
            {
               "type": "function",
               "function": "es",
               "arguments": [
                  {
                     "type": "literal",
                     "value": 404,
                  }
               ]
            }
         ]
      },
      {
         "type": "chain",
         "chain": [
            {
               "type": "function",
               "function": "es",
               "arguments": [
                  {
                     "type": "literal",
                     "value": 200,
                  }
               ]
            }
         ]
      }
   ]
 }
```

Now let's look at it in the new, commas required grammar. Again, autocompletion fluff removed. As we can see, there is only one chain, with each function as a link in it.. The interpreter will try to pass the output of `.es(404)` into `.es(200)`, which will fail.

```
{
   "tree": [
      {
         "type": "chain",
         "chain": [
            {
               "type": "function",
               "function": "es",
               "arguments": [
                  {
                     "type": "literal",
                     "value": 404
                  }
               ]
            },
            {
               "type": "function",
               "function": "es",
               "arguments": [
                  {
                     "type": "literal",
                     "value": 200
                  }
               ]
            }
         ]
      }
   ]
}
```

Cheers,
Rashid

rashidkpc commented on src/core_plugins/timelion/server/handlers/lib/__tests__/parse_sheet.js:
```diff
@@ -0,0 +1,33 @@
+  it('does not allow whitespace used between two expressions', function () {
+    const data = ['.es() .es(404)'];
+    expect(function () {
+      parseSheet(data);
```
This won't throw, `parseSheet` is only generating the AST. This statement generates an AST similar to the one below and is semantically valid.

ppisljar commented on src/core_plugins/timelion/server/handlers/__tests__/parse_sheet.js:
```diff
@@ -0,0 +1,50 @@
+        timezone: 'Europe/Berlin'
+      }
+    };
+    return await Promise.all(chainRunner.processRequest(request));
```
this does not work  .... it throws `object promise is not iterable` however the same works in `routes/run.js`
no idea whats wrong ... for some reason it doesn't wait here for the promises to resolve no matter what i try. help appreciated!

weltenwort replied:
The difference I see is that in `routes/run.js` the `Promise` is imported from bluebird while this uses `global.Promise`. Maybe there's a difference in the tolerance regarding being passed single promises vs arrays of `Promise.all` between bluebird and native promises. I'm not sure the `Promise.all` is required at all since `chainRunner.processRequest` always seems to return a single promise.

weltenwort replied:
I guess what we want here is

```javascript
return await Promise.resolve(request)
  .then(chainRunner.processRequest);
```

or using bluebird

```javascript
return await Promise.try(() => chainRunner.processRequest(request));
```

ppisljar replied:
thanks ... i somehow expected that its using the es6 native promise :(

rashidkpc replied:
This doesn't belong in this file, as it is testing the interpreter. You want to test that the grammar is generating a parser that is producing the expected AST

This will make the tests run faster, as well as make the tests more useful to the reader by conveying that we are only testing the parser.

This will also keep you from having to mock the `time` property.

epixa commented:
I removed the `blocker` label from this PR so it's not confused as a blocker for the 5.4.0 release.  I added a `blocked` label to the issue that is blocked on this PR.

rashidkpc commented on src/core_plugins/timelion/server/handlers/__tests__/parse_sheet.js:
```diff
@@ -0,0 +1,50 @@
+    const data = ['.es() .es(404)'];
+    expect(function () {
+      parse(data);
+    }).to.throw();
```
You shouldn't be calling up the interpreter to test this, rather get the properties on the AST. I explain the differences in the generated AST here: 

https://github.com/elastic/kibana/pull/10961#issuecomment-292534143

I should've been more specific in that comment. While interpreter tests would be nice the goal of this pull is only related to parsing and the tests should be testing the parser aspect.

tbragin commented:
Removed the 5.4.0 label, since this pull involves a breaking change in syntax (commas to be required, whereas previously they were optional). Seems like the kind of change we should be introducing in a major release, rather than minor. With 6.0 alphas around the corner, I'd recommend we target it to one of those.

thomasneirynck commented:
jenkins, test this

thomasneirynck commented on src/core_plugins/timelion/public/chain.peg:
```diff
@@ -21,7 +21,7 @@
 }
 
 start
- = tree:series+ {
```
any reason why this is removed? Don't entirely see the relation with the other changes.

thomasneirynck commented on src/core_plugins/timelion/server/handlers/__tests__/parse_sheet.js:
```diff
@@ -0,0 +1,23 @@
+
+const expect = require('chai').expect;
+
+describe('timelion parse_sheet function', function () {
```
we should add tests for the change in white-space behavior too (not just for the new comma-separated behavior).

ppisljar replied:
we do test for the white space behavior in the test below (it should not be allowed)

or do you mean testing that \n and \r are also accepted as white space ?

thomasneirynck replied:
yes, test that \n and \r are now accepted.

but not sure if we already have tests in place. if not, this could be a separate effort.

thomasneirynck commented on src/core_plugins/timelion/public/chain.peg:
```diff
@@ -132,10 +131,10 @@ reference
+ = func:function space? rest:function* {return {type: 'chain', chain: [func].concat(rest)}}
 
 group
- = '(' space? grouped:series+ space? ')' functions:function* {
```
dito

ppisljar replied:
dito :)

ppisljar commented:
@rashidkpc do you want to take another look ?

rashidkpc commented:
LGTM

rashidkpc commented:
A quick note: white space before the '.' will cause autocomplete to not trigger. Probably not a deal breaker as implementing the multi-line editor will require a reboot of the auto completion system anyway.

cjcenizal commented on src/core_plugins/timelion/server/handlers/__tests__/parse_sheet.js:
```diff
@@ -0,0 +1,31 @@
+    expect(ast[0][0].type).to.equal('chain');
+    expect(ast[0][1].type).to.equal('chain');
+  });
+});
```
I was a little confused by the use of "allow" and "does not allow" in the test descriptions, because I assumed we'd be throwing errors or somehow preventing disallowed input. After I dug into the tests, I realized this was an incorrect assumption. I think it will be a little clearer if we describe the behavior as either "splitting expressions" or "not splitting expressions". Thoughts?

Because I'm not familiar with the structure of the generated AST, I also found it useful to break out the part of the object that represents the list of expressions into a variable, to increase readability.

```javascript
const parseSheet = require('../lib/parse_sheet');

const expect = require('chai').expect;

describe('timelion parse_sheet function', function () {
  it(`doesn't split expressions on whitespace`, async function () {
    const data = ['.es() .es(404)'];
    const ast = parseSheet(data);

    const expressions = ast[0];
    expect(expressions.length).to.equal(1);
    expect(expressions[0].type).to.equal('chain');
  });]

  it('splits expressions on commas', function () {
    const data = ['.es(), .es(404)'];
    const ast = parseSheet(data);

    const expressions = ast[0];
    expect(expressions.length).to.equal(2);
    expect(expressions[0].type).to.equal('chain');
    expect(expressions[1].type).to.equal('chain');
  });

  it('splits expressions on newlines', function () {
    const data = [`.es()\n\r ,\n\r .es(404)`];
    const ast = parseSheet(data);

    const expressions = ast[0];
    expect(expressions.length).to.equal(2);
    expect(expressions[0].type).to.equal('chain');
    expect(expressions[1].type).to.equal('chain');
  });
});
```

epixa commented:
@ppisljar I'm not sure where this landed - is it still a breaking change?  If so, can you please update the breaking change docs for 6.0 with details?

ppisljar commented:
@epixa thanks for the reminder! it landed in 6.0
